packages/math-input/src/components/key-translator.ts

```diff
@@ -0,0 +1,182 @@
+    return function (mathQuill: MathQuillInterface) {
+        switch (type) {
+            case ActionType.WRITE: {
+                mathQuill.write(str);
+                return;
+            }
+            case ActionType.CMD: {
+                mathQuill.cmd(str);
+                return;
+            }
+            case ActionType.KEYSTROKE: {
+                mathQuill.keystroke(str);
+                return;
+            }
+        }
+    };
+}
+
+const keyToMathquillMap: Record<Key, MathQuillCallback | null> = {
+    CDOT: getGenericCallback("\\cdot"),
+    COS: getGenericCallback("cos"),
+    DECIMAL: getGenericCallback("."),
+    DIVIDE: getGenericCallback("\\div"),
+    EQUAL: getGenericCallback("="),
+    EXP: getGenericCallback("^"),
+    EXP_2: getGenericCallback("^2"),
+    EXP_3: getGenericCallback("^3"),
+    GEQ: getGenericCallback("\\geq"),
+    GT: getGenericCallback(">"),
+    LEQ: getGenericCallback("\\leq"),
+    LN: getGenericCallback("\\ln"),
+    LOG: getGenericCallback("\\log"),
+    LOG_N: getGenericCallback("log_{ }"),
+    LT: getGenericCallback("<"),
+    MINUS: getGenericCallback("-"),
+    NEGATIVE: getGenericCallback("-"),
+    NEQ: getGenericCallback("\\neq"),
+    PERCENT: getGenericCallback("%"),
+    PERIOD: getGenericCallback("."),
+    PLUS: getGenericCallback("+"),
+    SIN: getGenericCallback("sin"),
+    TAN: getGenericCallback("tan"),
+    TIMES: getGenericCallback("\\times"),
+
+    // The `FRAC_EXCLUSIVE` variant is handled manually, since we may need to do
+    // some additional navigation depending on the cursor position.
+    FRAC_INCLUSIVE: getGenericCallback("/", ActionType.CMD),
+    LEFT_PAREN: getGenericCallback("(", ActionType.CMD),
+    RIGHT_PAREN: getGenericCallback(")", ActionType.CMD),
+    SQRT: getGenericCallback("sqrt", ActionType.CMD),
+    PHI: getGenericCallback("\\phi", ActionType.CMD),
+    PI: getGenericCallback("pi", ActionType.CMD),
+    THETA: getGenericCallback("theta", ActionType.CMD),
+    RADICAL: getGenericCallback("nthroot", ActionType.CMD),
+
+    UP: getGenericCallback("Up", ActionType.KEYSTROKE),
+    DOWN: getGenericCallback("Down", ActionType.KEYSTROKE),
+
+    LOG_B: (mathQuill) => {
+        mathQuill.typedText("log_");
+        mathQuill.keystroke("Right");
+        mathQuill.typedText("(");
+        mathQuill.keystroke("Left");
+        mathQuill.keystroke("Left");
+    },
+
+    NTHROOT3: (mathQuill) => {
+        mathQuill.typedText("nthroot3");
+        mathQuill.keystroke("Right");
+    },
+
+    POW: (mathQuill) => {
+        const contents = mathQuill.latex();
+        mathQuill.typedText("^");
+
+        // If the input hasn't changed (for example, if we're
+        // attempting to add an exponent on an empty input or an empty
+        // denominator), insert our own "a^b"
+        if (mathQuill.latex() === contents) {
+            mathQuill.typedText("a^b");
+        }
+    },
+
+    // These need to be overwritten by the consumer
```

Comment:
The consumer being... the app that is integrating this translator?  

Does this change per integration?  Could we lift those translations up into here?

Reply:
Right, we might expect different behavior depending on the context (such as mobile vs desktop vs Khanmigo). Like for instance Perseus' MathInput and MathInput's MathInput treat "FRAC" differently.

I'm trying to not editorialize too much right now, just colocate duplicated logic. But because I used Key for the type, it wanted me to have something for every option.